src/Processors/Formats/Impl/CSVRowInputFormat.cpp

```diff
@@ -315,14 +316,48 @@ bool CSVFormatReader::readField(
         return false;
     }
 
-    if (format_settings.null_as_default && !isNullableOrLowCardinalityNullable(type))
+    BufferBase::Position pos_start = buf->position();
```

Comment:
The idea is ok, but the implementation has several problems:
1) We cannot just remember current position and then reset buffer back to it, because internal buffer can be refilled with new data and `pos_start` can point to some invalid data. And in general I don't understand why do we need to reset position at all. readCSVField should be enough, no need to reset position back and skip field (otherwise we just do same work twice).
2) Throwing exception and catching it is bad practice and in general `LOGICAL_ERROR` is severe error and should not be used like this.
3) Code duplication of serialization.

I suggest splitting the logic and adding 2 new mehthods `readFieldImpl` and `readFieldOrDefault`.
So, in `readField` we will do just
```c++
...
if (format_settings.csv.use_default_on_bad_values)
    return readFieldOrDefault(column, type, serialization);
return readFieldImpl(*buf, column, type, serialization);
```

The implementation of new methods can look like this (I didn't test it, so it can be not 100% correct, please, check it)
```c++
bool CSVFormatReader::readFieldImpl(ReadBuffer & istr, DB::IColumn & column, const DB::DataTypePtr & type, const DB::SerializationPtr & serialization)
{
    if (format_settings.null_as_default && !isNullableOrLowCardinalityNullable(type))
    {
        /// If value is null but type is not nullable then use default value instead.
        return SerializationNullable::deserializeTextCSVImpl(column, istr, format_settings, serialization);
    }

    /// Read the column normally.
    serialization->deserializeTextCSV(column, istr, format_settings);
    return true;
}

bool CSVFormatReader::readFieldOrDefault(DB::IColumn & column, const DB::DataTypePtr & type, const DB::SerializationPtr & serialization)
{
    String field;
    readCSVField(field, *buf, format_settings.csv);
    ReadBufferFromString tmp_buf(field);
    bool is_bad_value = false;
    bool res = false;

    size_t col_size = column.size();
    try
    {
        res = readFieldImpl(tmp_buf, column, type, serialization);
        /// Check if we parsed the whole field successfully.
        if (!field.empty() && !tmp_buf.eof())
            is_bad_value = true;
    }
    catch (const Exception &)
    {
        is_bad_value = true;
    }

    if (!is_bad_value)
        return res;

    if (column.size() == col_size + 1)
        column.popBack(1);
    column.insertDefault();
    return false;
}
```

Reply:
done.  I have test these code on my suitation,  and it works. thanks for your advices, and please reivew the code again. @Avogar